common/src/test/java/org/conscrypt/BufferUtilsTest.java

```diff
@@ -0,0 +1,202 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.conscrypt;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertTrue;
+
+import java.nio.ByteBuffer;
+import org.conscrypt.TestUtils.BufferType;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+
+@RunWith(Parameterized.class)
+public class BufferUtilsTest {
+    private static final int K64 = 64 * 1024;
+    private static final int K16 = 16 * 1024;
+
+    private static final int[][] TEST_SIZES = {
+            // All even numbers as several tests use size/2
+            { 0 },
+            { 0, 0, 0, 0 },
+            { 0, 0, 0, 2 },
+            { 2, 0, 0, 0 },
+            { 100, 200, 300 },
+            { 1000, 2000, 3000 },
+            { K16 },
+            { 0, 0, K16 },
+            { K16, 0, 0 },
+            { K64 },
+            { 0, 0, K64 },
+            { K64, 0, 0 },
+            { 100, 100, K64 },
+            { K64, 100, 100 },
+            { K64, K64, K64 },
+    };
+
+
+    @Parameters(name = "{0}")
+    public static BufferType[] data() {
+        return new BufferType[] { BufferType.HEAP, BufferType.DIRECT };
+    }
+
+    @Parameter
+    public BufferType bufferType;
+
+    @Test
+    public void noNulls() {
+        for (int[] sizes : TEST_SIZES) {
+            assertTrue(BufferUtils.noNulls(bufferType.newRandomBuffers(sizes)));
+        }
+
+        ByteBuffer[] buffers = bufferType.newRandomBuffers(10, 10, 10, 10, 10);
+        buffers[2] = null;
+        assertFalse(BufferUtils.noNulls(buffers));
+    }
+
+    @Test
+    public void remaining() {
+        for (int[] sizes : TEST_SIZES) {
+            assertEquals(arraySum(sizes),
+                    BufferUtils.remaining(bufferType.newRandomBuffers(sizes)));
+        }
+    }
+
+    @Test
+    public void consume() {
+        for (int[] sizes : TEST_SIZES) {
+            ByteBuffer[] buffers = bufferType.newRandomBuffers(sizes);
+            int totalSize = arraySum(sizes);
+
+            BufferUtils.consume(buffers, 0);
+            assertEquals(totalSize, BufferUtils.remaining(buffers));
+
+            BufferUtils.consume(buffers,totalSize / 2);
+            assertEquals(totalSize / 2, BufferUtils.remaining(buffers));
+
+            BufferUtils.consume(buffers,totalSize / 2);
+            assertEquals(0, BufferUtils.remaining(buffers));
+
+            try {
+                BufferUtils.consume(buffers, totalSize / 2);
```

Comment:
Please add fail();

Reply:
Oops, thanks! Done.

Reply:
And that was masking a bug in the test ... thanks! :)